include/cadmium/basic_model/iestream.hpp

```diff
@@ -0,0 +1,188 @@
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __IESTREAM_HPP__
+#define __IESTREAM_HPP__
+
+#include <cadmium/modeling/ports.hpp>
+#include <cadmium/modeling/message_bag.hpp>
+#include <limits>
+
+#include <stddef.h>
+
+#include <string>
+#include <fstream>
+
+using namespace std;
+namespace cadmium::basic_models::pdevs {
+
+/**
+ * @brief input_event_stream PDEVS Model.
+ *
+ * iestream PDEVS Model plays a history of events received by an input stream.
+ * The list of events allows to be used as a connector to external tools.
+ * The input format is "time output and a custom parser can be defined."
+ * Data type MSG must have the operator >> in order to work. 
+ * Data type TIME must also have the operator >> in order to work.
+ * Each line must be an MSG. Therefore the operator >> cannot read inputs that are specified in multiple lines
+ *
+*/
+
+
+template<class TIME, class INPUT>
+class Parser {
+    private:
+        std::ifstream file;
+    public:
+        // Constructors
+        Parser() {}
+        Parser(const char* file_path) {
+            this->open_file(file_path);
+        }
+        void open_file(const char* file_path) {
+            file.open(file_path);
+        }
+    std::pair<TIME,INPUT> next_timed_input() {
+        INPUT result;
+        TIME next_time;
+        if (file.eof()) throw std::exception();
+        file >> next_time;
+        file >> result;
+        return std::make_pair(next_time,result);
+    }
+};
+
+template<typename MSG> struct iestream_input_defs{
+    //custom ports
+    struct out : public out_port<MSG> {};
+};
+
+template<typename MSG, typename TIME>
+class iestream_input {
+    using defs=iestream_input_defs<MSG>;
+    public:
+        // default constructor
+        iestream_input() noexcept {}
+        iestream_input(const char* file_path) noexcept {
+            state._parser.open_file(file_path);
+        }
+    // state definition
+    struct state_type{
+        Parser<TIME, MSG> _parser;
+        MSG _last_input_read;
+        vector<MSG> _next_input;
+        TIME _simulation_time = TIME();
+        TIME _next_time = TIME();
+        TIME _next_time2 = TIME();
+        bool _initialization = true;
+    }; 
+    state_type state;
+
+    // ports definition
+    using input_ports=std::tuple<>;
+    using output_ports=std::tuple<typename defs::out>;
+
+    // internal transition
+    void internal_transition() {
+        //out << "INTERNAL. IESTREAM " << endl;
```

Comment:
You can pass this to the logger to keep track of it. We can add a it to the info level or debug level easily.

Reply:
It was an old comment for debugging. Removed